File: ImperatorToCK3Tests/CommonUtilitiesTests/DatedHistoryBlockTests.cpp
#include "gtest/gtest.h"
#include "CommonUtilities/DatedHistoryBlock.h"
#include <sstream>



TEST(CommonUtilities_DatedHistoryBlockTests, onlyRegisteredPairsAreRead) {
	std::stringstream input;
	input << R"( = {
					culture = cuman
					culture = bashkiri
					religion = jewish
					title = "c_sarkel"
					development = 500
					monthly_alien_sightings = 5
				})";

	auto datedHistoryBlock = DatedHistoryBlock{
		{
			{ "culture", "culture", ""},
			{ "title", "title", ""},
			{ "monthly_alien_sightings", "monthly_alien_sightings", "0" }
		},
		{},
		input
	};
	auto contents = datedHistoryBlock.getContents();

	ASSERT_EQ(3, contents.simpleFieldContents.size());

	ASSERT_EQ("cuman", contents.simpleFieldContents.at("culture")[0]);
	ASSERT_EQ("bashkiri", contents.simpleFieldContents.at("culture")[1]);
	ASSERT_EQ("c_sarkel", contents.simpleFieldContents.at("title").back());
	ASSERT_EQ("5", contents.simpleFieldContents.at("monthly_alien_sightings").back());

	ASSERT_FALSE(contents.simpleFieldContents.contains("religion"));
	ASSERT_FALSE(contents.simpleFieldContents.contains("development"));
}


TEST(CommonUtilities_DatedHistoryBlockTests, quotedStringsAreNotReadAsKeys) {
	std::stringstream input;
	input << R"( = {
					culture = cuman
					"religion" = jewish
				})";

	const auto& contents = DatedHistoryBlock{ { { "religion", "religion", std::nullopt } }, {}, input }.getContents();

	ASSERT_EQ(1, contents.simpleFieldContents.size());
	ASSERT_FALSE(contents.simpleFieldContents.contains(R"("religion")"));
}

File: ImperatorToCK3.UnitTests/CommonUtils/DatedHistoryBlockTests.cs
using commonItems;
using ImperatorToCK3.CommonUtils;
using System.Collections.Generic;
using System.Linq;
using Xunit;

namespace ImperatorToCK3.UnitTests.CommonUtils {
	public class DatedHistoryBlockTests {
		[Fact]
		public void OnlyRegisteredThingsAreReturned() {
			var reader = new BufferedReader(
				@" = {
				culture = cuman
				culture = bashkiri
				religion = jewish
				title = ""c_sarkel""
				development = 500
				monthly_alien_sightings = 5
			}");

			var datedHistoryBlock = new DatedHistoryBlock(
				new List<SimpleFieldDef>
				{
					new() {FieldName = "culture", Setters = {"culture"}, InitialValue = ""},
					new() {FieldName = "title", Setters = {"title"}, InitialValue = ""},
					new() {
						FieldName = "monthly_alien_sightings", Setters = {"monthly_alien_sightings"}, InitialValue = "0"
					}
				},
				new List<ContainerFieldDef>(),
				reader
			);

			Assert.Equal(3, datedHistoryBlock.SimpleFieldContents.Count);

			Assert.Equal("cuman", datedHistoryBlock.SimpleFieldContents["culture"][0].Value);
			Assert.Equal("bashkiri", datedHistoryBlock.SimpleFieldContents["culture"][1].Value);
			Assert.Equal("c_sarkel", datedHistoryBlock.SimpleFieldContents["title"].Last().Value);
			Assert.Equal(5, datedHistoryBlock.SimpleFieldContents["monthly_alien_sightings"].Last().Value);

			Assert.False(datedHistoryBlock.SimpleFieldContents.ContainsKey("religion"));
			Assert.False(datedHistoryBlock.SimpleFieldContents.ContainsKey("development"));
		}

		[Fact]
		public void QuotedStringsAreNotReadAsKeys() {
			var reader = new BufferedReader(
				@"= {
					culture = cuman
					""religion"" = jewish
				}"
			);

			var datedHistoryBlock = new DatedHistoryBlock(
				new List<SimpleFieldDef> {
					new() { FieldName = "religion", Setters = {"religion"}, InitialValue = null }
				},
				new List<ContainerFieldDef>(),
				reader
			);

			Assert.Empty(datedHistoryBlock.SimpleFieldContents);
		}
	}
}
